pytest_container/pod.py

```diff
@@ -0,0 +1,217 @@
+
+
+@dataclass(frozen=True)
+class PodData:
+    """Class that is returned by the ``pod`` and ``pod_per_test`` fixtures. It
+    contains all necessary information about the created pod and the containers
+    running inside it.
+
+    """
+
+    #: containers that are in the pod
+    containers: List[Union[DerivedContainer, Container]]
+
+    #: The :py:class:`~pytest_container.container.ContainerData` instances of
+    #: each container in the pod.
+    container_data: List[ContainerData]
+
+    #: unique id/hash of the running pod
+    pod_id: str
+
+    #: unique id/hash of the infra container of the pod
+    infra_container_id: str
+
+    #: ports exposed by this pod
+    forwarded_ports: List[PortForwarding]
+
+
+@dataclass
+class PodLauncher:
+    """A context manager that creates, starts and destroys a pod with all of its
+    containers.
+
+    """
+
+    #: the pod that should be created and launched including all of its
+    #: containers
+    pod: Pod
+
+    #: root directory of the pytest testsuite
+    rootdir: Path
+
+    #: optional name of the pod
+    pod_name: str = ""
+
+    #: additional arguments to pass to the container build commands
+    extra_build_args: List[str] = field(default_factory=list)
+
+    #: additional arguments to pass to the container run commands
+    extra_run_args: List[str] = field(default_factory=list)
+
+    #: additional arguments to pass to the ``pod create`` command
+    extra_pod_create_args: List[str] = field(default_factory=list)
+
+    _launchers: List[ContainerLauncher] = field(default_factory=list)
+    _new_port_forwards: List[PortForwarding] = field(default_factory=list)
+
+    #: id of the pod
+    _pod_id: Optional[str] = None
+
+    #: id of the infra pod
+    _infra_container_id: Optional[str] = None
+
+    _stack: contextlib.ExitStack = field(default_factory=contextlib.ExitStack)
+
+    def __enter__(self) -> "PodLauncher":
+        runtime = get_selected_runtime()
+        if runtime != PodmanRuntime():
+            raise RuntimeError(
+                f"pods can only be created with podman, but got {runtime}"
+            )
+
+        create_cmd = [runtime.runner_binary, "pod", "create"] + (
+            ["--name", self.pod_name] if self.pod_name else []
+        )
+
+        if self.pod.port_forwardings:
+            with FileLock(self.rootdir / "port_check.lock"):
+                self._new_port_forwards = create_host_port_port_forward(
+                    self.pod.port_forwardings
+                )
+                # port_forward_args = []
+                for new_forward in self._new_port_forwards:
+                    create_cmd += new_forward.forward_cli_args
+
+                _logger.debug("Creating pod via: %s", create_cmd)
+                self._pod_id = check_output(create_cmd).decode().strip()
+        else:
+            _logger.debug("Creating pod via: %s", create_cmd)
+            self._pod_id = check_output(create_cmd).decode().strip()
```

Comment:
is there's an explicit else here because you're holding the lock here? otherwise seems to be a redundant else, can we just mutate `create_cmd` based on whether port fwding is required or not and then create the pod just once?
```suggestion
        _logger.debug("Creating pod via: %s", create_cmd)
        self._pod_id = check_output(create_cmd).decode().strip()
```

Reply:
So this code block is copy-pasta from `container.py` with iirc a better comment explaining the locking. But the tl;dr; is: we use this lock so that we look for free ports on the host from a single process only. But that creates a bottleneck for launching containers and so we don't grab the lock if no port forwarding is necessary. Also, we need to hold the lock until the container with the forwarded ports is started. Otherwise you run into nasty issues like: you grab the lock, you find a free port, you release the lock. Then immediately another containers is scheduled to launch and starts looking for free ports, grabs the one that "belongs" to the previous container because it hasn't launched and thus blocked the port before.

Hope this makes sense… The code is unfortunately still not 100% robust, but it's the best I've managed to create so far.

Reply:
Thanks, that makes sense.